// @helia/routers/dist/src/delegated-http-routing.js
import { createDelegatedRoutingV1HttpApiClient } from '@helia/delegated-routing-v1-http-api-client';
import { CodeError } from '@libp2p/interface';
import { marshal, unmarshal, peerIdFromRoutingKey } from 'ipns';
import first from 'it-first';
import map from 'it-map';
import { equals as uint8ArrayEquals } from 'uint8arrays/equals';
import { fromString as uint8ArrayFromString } from 'uint8arrays/from-string';
const IPNS_PREFIX = uint8ArrayFromString('/ipns/');
function isIPNSKey(key) {
    return uint8ArrayEquals(key.subarray(0, IPNS_PREFIX.byteLength), IPNS_PREFIX);
}
class DelegatedHTTPRouter {
    client;
    constructor(url) {
        this.client = createDelegatedRoutingV1HttpApiClient(url);
    }
    async provide(cid, options) {
        // noop
    }
    async *findProviders(cid, options) {
        yield* map(this.client.getProviders(cid, options), (record) => {
            return {
                id: record.ID,
                multiaddrs: record.Addrs,
                protocols: record.Protocols
            };
        });
    }
    async put(key, value, options) {
        if (!isIPNSKey(key)) {
            return;
        }
        const peerId = peerIdFromRoutingKey(key);
        const record = unmarshal(value);
        await this.client.putIPNS(peerId, record, options);
    }
    async get(key, options) {
        if (!isIPNSKey(key)) {
            throw new CodeError('Not found', 'ERR_NOT_FOUND');
        }
        const peerId = peerIdFromRoutingKey(key);
        try {
            const record = await this.client.getIPNS(peerId, options);
            return marshal(record);
        }
        catch (err) {
            // ERR_BAD_RESPONSE is thrown when the response had no body, which means
            // the record couldn't be found
            if (err.code === 'ERR_BAD_RESPONSE') {
                throw new CodeError('Not found', 'ERR_NOT_FOUND');
            }
            throw err;
        }
    }
    async findPeer(peerId, options) {
        const peer = await first(this.client.getPeers(peerId, options));
        if (peer != null) {
            return {
                id: peer.ID,
                multiaddrs: peer.Addrs ?? []
            };
        }
        throw new CodeError('Not found', 'ERR_NOT_FOUND');
    }
    async *getClosestPeers(key, options) {
        // noop
    }
}
/**
 * Creates a Helia Router that connects to an endpoint that supports the [Delegated Routing V1 HTTP API](https://specs.ipfs.tech/routing/http-routing-v1/) spec.
 */
export function delegatedHTTPRouting(url) {
    return new DelegatedHTTPRouter(new URL(url));
}
