// @libp2p/webrtc/dist/src/maconn.js
import { nopSink, nopSource } from './util.js';
export class WebRTCMultiaddrConnection {
    log;
    /**
     * WebRTC Peer Connection
     */
    peerConnection;
    /**
     * The multiaddr address used to communicate with the remote peer
     */
    remoteAddr;
    /**
     * Holds the lifecycle times of the connection
     */
    timeline;
    /**
     * Optional metrics counter group for this connection
     */
    metrics;
    /**
     * The stream source, a no-op as the transport natively supports multiplexing
     */
    source = nopSource();
    /**
     * The stream destination, a no-op as the transport natively supports multiplexing
     */
    sink = nopSink;
    constructor(components, init) {
        this.log = components.logger.forComponent('libp2p:webrtc:maconn');
        this.remoteAddr = init.remoteAddr;
        this.timeline = init.timeline;
        this.peerConnection = init.peerConnection;
        const initialState = this.peerConnection.connectionState;
        this.peerConnection.onconnectionstatechange = () => {
            this.log.trace('peer connection state change', this.peerConnection.connectionState, 'initial state', initialState);
            if (this.peerConnection.connectionState === 'disconnected' || this.peerConnection.connectionState === 'failed' || this.peerConnection.connectionState === 'closed') {
                // nothing else to do but close the connection
                this.timeline.close = Date.now();
            }
        };
    }
    async close(options) {
        this.log.trace('closing connection');
        this.peerConnection.close();
        this.timeline.close = Date.now();
        this.metrics?.increment({ close: true });
    }
    abort(err) {
        this.log.error('closing connection due to error', err);
        this.peerConnection.close();
        this.timeline.close = Date.now();
        this.metrics?.increment({ abort: true });
    }
}
